### PR TITLE
RegexArrayShapeMatcher: Support resolving of constants in patterns

### PR DESCRIPTION
**Pull request by staabm**
Refactors RegexArrayShapeMatcher to re-use InitializerExprTypeResolver

closes https://github.com/phpstan/phpstan/issues/11384

### CODE REVIEW

**Comment by staabm** [comment]
I think I now have a better understand how the `InitializerExprTypeResolver` works and I will have a look whether we can do a similar thing within `RegularExpressionQuotingRule`

**Comment by Seldaek** [comment]
Cool, less code and less bugs :D

File: src/Type/Php/RegexArrayShapeMatcher.php [hunk 1]
```diff
@@ -724,38 +726,33 @@ private function getPatternType(Expr $patternExpr, Scope $scope): Type
...
-			&& $concat->right->name instanceof Name
-			&& $concat->right->name->toLowerString() === 'preg_quote'
-		) {
-			$right = new ConstantStringType('');
-		} elseif ($concat->right instanceof Expr\BinaryOp\Concat) {
-			$right = $this->resolvePatternConcat($concat->right, $scope);
-		} else {
-			$right = $scope->getType($concat->right);
-		}
+		$resolveConcat = static function (Expr $expr) use (&$resolveConcat, $scope) {
```
Comment (ondrejmirtes) [inline src/Type/Php/RegexArrayShapeMatcher.php]: These are usually memory leaks. It'd be nicer to have this as a private method.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/RegexArrayShapeMatcher.php (hunk 1 above)
Comment (staabm) [reply src/Type/Php/RegexArrayShapeMatcher.php]: I went with a inline class, hopefully that works for you.
didn't have a better idea on how to get the lexical $scope variable into the callable

File: src/Type/Php/RegexArrayShapeMatcher.php [hunk 2]
```diff
@@ -724,38 +726,42 @@ private function getPatternType(Expr $patternExpr, Scope $scope): Type
...
+				}
+
+				return $this->scope->getType($expr);
 			}
-		}
 
-		return TypeCombinator::union(...$strings);
+		};
+
+		return $this->initializerExprTypeResolver->getConcatType($concat->left, $concat->right, static fn (Expr $expr): Type => $resolver->resolve($expr));
```
Comment (Seldaek) [inline src/Type/Php/RegexArrayShapeMatcher.php:764]: Not sure if this is supported by the php8.1=>7.2 transpile but hopefully :)

```suggestion
		return $this->initializerExprTypeResolver->getConcatType($concat->left, $concat->right, $resolver->resolve(...));
```

Comment (ondrejmirtes) [reply src/Type/Php/RegexArrayShapeMatcher.php:764]: It's not :)

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**